R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
+
+    # If the input parameters are measured in different units
+    # but are convertible from one to another (and this kind of conversion supported)
+    # then modify the formula in order to perform units conversion on the fly
+
+    if (param_units[[denominator_code]] != param_units[[numerator_code]]) {
+      # Find conversion factor for denominator
+      conv_factor <- get_conv_factor(
+        param_units[[denominator_code]],
+        param_units[[numerator_code]]
+      )
+
+      if (!is.na(conv_factor)) {
+        ratio_formula <- expr(
+          !!sym(paste0("AVAL.", numerator_code)) / (
+            !!sym(paste0("AVAL.", denominator_code)) * !!conv_factor
+          )
+        )
+
+        cli_alert_info(
+          "Unit conversion performed for {.val {denominator_code}}. Values converted from
+          {.val {param_units[[denominator_code]]}} to {.val {param_units[[numerator_code]]}}.",
+          wrap = TRUE
+        )
+      }
+    }
+  }
+
+  ### Identify constant parameters ----
+
+  parameters <- c(numerator_code, denominator_code)
+  constant_parameters <- NULL
+
+  if (constant_numerator) {
+    constant_parameters <- c(constant_parameters, numerator_code)
+
+    parameters <- parameters %>%
+      .[!. == numerator_code]
+  }
+
+  if (constant_denominator) {
+    constant_parameters <- c(constant_parameters, denominator_code)
+
+    parameters <- parameters %>%
+      .[!. == denominator_code] %>%
+      ifelse(length(.) == 0, NULL, .)
+  }
+
+  ### Call the core {admiral} function to derive Ratio parameter ----
+
+  derive_param_computed(
+    dataset,
+    filter = !!filter,
+    parameters = parameters,
+    by_vars = by_vars,
+    set_values_to = exprs(
+      AVAL = !!ratio_formula,
+      !!!set_values_to
+    ),
+    constant_parameters = constant_parameters,
+    constant_by_vars = constant_by_vars
+  )
+}
+
+#' Unit conversion
+#'
+#' @name unit-conversion
+#' @keywords internal
+NULL
+#> NULL
+
+#' @description `get_conv_factor()` extracts a conversion factor for a pair of units.
+#' Returns `NA` if units are not supported/convertible.
+#'
+#' @rdname unit-conversion
+#' @keywords internal
+get_conv_factor <- function(from_unit, to_unit) {
+  # Get all conversion factors supported
+  conv_factors_all <- get_conv_factors_all()
+
+  # Return conversion factor if units are supported and convertible
+  for (unit_category in names(conv_factors_all)) {
+    if (all(c(from_unit, to_unit) %in% names(conv_factors_all[[unit_category]]))) {
+      return(
+        conv_factors_all[[unit_category]][[from_unit]] /
+          conv_factors_all[[unit_category]][[to_unit]]
+      )
+    }
+  }
+
+  # If units are not supported/convertible
+  return(NA_real_)
+}
+
+#' @description `get_conv_factors_all()` returns all conversion factors supported.
+#'
+#' **Note:** Conversion factor for inch comes from the NCI Thesaurus
```

Comment:
Updated to

`**Note:** Conversion factors comes from unit definitions as per the NCI Thesaurus`

Reply:
I would actually prefer to write the actual conversions/factors, even though it will be verbose. I.e. "A meter is defined as 100cm". That would make it easier for users to quickly check that the unit conversions make sense. I also think we should refer to the CDISC definitions and not NCI, even though CDISC just copy/paste the NCI definition.

Reply:
Added

![image](https://github.com/user-attachments/assets/85d4f2c9-5eee-4c64-ad6f-760f8b7da7f9)